### ISSUE TITLE
[css-scroll-anchoring-1] Anchor selection algorithm can lead to random results

### ISSUE THREAD

**Issue by bzbarsky** [w3c/csswg-drafts#4124]
The way that https://drafts.csswg.org/css-scroll-anchoring/#anchoring-algorithm works is:

1) Walk the child nodes of our scrollable thing.
2) Stop when we find one that meets certain criteria.

The thing is, the algorithm doesn't say what order the child nodes are walked in.  As far as I can tell, you can walk them forwards, backwards, or in any random order, and those would all be spec-compliant.

Note that just walking in DOM order from start to end may not be the right thing either; e.g. if the parent is a grid the earlier DOM kids may not be the thing you want to anchor at because they are further from the block start edge than later kids are.

@emilio @skobes

**Comment by emilio** [comment]
Yeah, in practice https://github.com/w3c/csswg-drafts/issues/3480 was also about this, since DOM order doesn't include any shadow things. And in practice both implementations walk the layout tree. Yay?

**Comment by bzbarsky** [comment]
Walking the layout tree actually makes sense, at least... ;)  Now if only it were defined somewhere.

**Comment by zcorpan** [comment]
Is defining this in terms of the flat tree what we want? Or is layout tree what is really needed? Is it fully documented somewhere how these trees differ?

**Comment by emilio** [comment]
Yes, the flat tree includes display: contents elements for example, which don't appear in the layout tree. The layout tree also contains all sorts of anonymous boxes and such, though those are excluded from the candidate examination algorithm in implementations.
